### PR TITLE
*: remove namespace

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve? 
See #1416.

### What is changed and how it works?
This PR removes all namespace related logic since it's barely used and complexity. Specifically,
1. remove `namespace` package and `table` package.
2. move codec related logic into `pkg/codec`
3. update the related docs

### Check List 

Tests 

 - Unit test

Code changes

 - Has configuration change
 - Has HTTP API interfaces change
 - Has persistent data change

Side effects

 - Breaking backward compatibility (for someone who use table classifier may need to change to the `enable-global-merge` option)

Related changes

 - Need to update the documentation

### CODE REVIEW

File: conf/config.toml [hunk 1]
```diff
@@ -68,28 +66,36 @@ region-schedule-limit = 64
 replica-schedule-limit = 64
 merge-schedule-limit = 8
 hot-region-schedule-limit = 4
-#tolerant-size-ratio = 0.0
-#enable-one-way-merge = false
-
-# customized schedulers, the format is as below
-# if empty, it will use balance-leader, balance-region, hot-region as default
+## When the score difference between the leader or Region of the two stores is
```
Comment (lhy1024) [inline conf/config.toml]: Do we need to replace "Region" with "region"?

File: server/schedule/checker/merge_checker.go [hunk 2]
```diff
@@ -134,8 +132,12 @@ func (m *MergeChecker) Check(region *core.RegionInfo) []*operator.Operator {
 }
 
 func (m *MergeChecker) checkTarget(region, adjacent *core.RegionInfo) bool {
+	if !m.cluster.IsGlobalMergeEnabled() {
```
Comment (lhy1024) [inline server/schedule/checker/merge_checker.go]: we need to check if `adjacent` is nil. I think it is better to abstracte it into a function `allowMerge` in `MergeChecker`.

Comment (lhy1024) [inline server/schedule/checker/merge_checker.go]: Do we neet to add test for `IsGlobalMergeEnabled`?

File: tools/pd-ctl/README.md [hunk 3]
```diff
@@ -203,6 +196,12 @@ Usage:
     >> config set enable-one-way-merge true  // Enable one way merge.
     ```
 
+- `enable-global-merge` controls the merge scheduler behavior. This means two regions belongs to the different tables can be merged.
```
Comment (lhy1024) [inline tools/pd-ctl/README.md]:
```suggestion
- `enable-global-merge` controls the merge scheduler behavior. This means two regions belong to the different tables can be merged.
```

File: server/config/config.go [hunk 4]
```diff
@@ -488,6 +476,12 @@ type ScheduleConfig struct {
 	SplitMergeInterval typeutil.Duration `toml:"split-merge-interval,omitempty" json:"split-merge-interval"`
 	// EnableOneWayMerge is the option to enable one way merge. This means a Region can only be merged into the next region of it.
 	EnableOneWayMerge bool `toml:"enable-one-way-merge,omitempty" json:"enable-one-way-merge,string"`
+	// EnableCrossTableMerge is the option to enable cross table merge. This means two Regions can be merged with different table IDs.
+	// This option only works when merge strategy is "table".
+	EnableCrossTableMerge bool `toml:"enable-cross-table-merge,omitempty" json:"enable-cross-table-merge,string"`
+	// MergeStrategy is option to control the merge strategy.
+	// There are some strategics supported: ["table", "raw", "txn"], default: "table"
+	MergeStrategy string `toml:"merge-strategy" json:"merge-strategy,string"`
```
Comment (disksing) [inline server/config/config.go]: em, maybe something like `cluster-type`? and move it to uppder level? (in Config not in ScheduleConfig)

File: server/schedule/checker/merge_checker.go [hunk 5]
```diff
@@ -134,8 +132,49 @@ func (m *MergeChecker) Check(region *core.RegionInfo) []*operator.Operator {
...
+	case "table":
+		allow := m.cluster.IsCrossTableMergeEnabled()
+		if allow {
+			return checkTable(region, adjacent)
+		}
+		return checkTableSame(region, adjacent)
+	case "raw":
+		return true
+	case "txn":
+		return checkTxnKeys(region, adjacent)
```
Comment (disksing) [inline server/schedule/checker/merge_checker.go]: I think can return `true` too.

File: server/schedule/checker/merge_checker.go [hunk 6]
```diff
@@ -134,8 +132,49 @@ func (m *MergeChecker) Check(region *core.RegionInfo) []*operator.Operator {
...
+// allowMerge returns true if two regions can be merged according to the merge strategy.
+func (m *MergeChecker) allowMerge(region *core.RegionInfo, adjacent *core.RegionInfo) bool {
+	strategy := m.cluster.GetMergeStrategy()
+	switch strategy {
+	case "table":
+		allow := m.cluster.IsCrossTableMergeEnabled()
+		if allow {
+			return checkTable(region, adjacent)
+		}
+		return checkTableSame(region, adjacent)
```
Comment (disksing) [inline server/schedule/checker/merge_checker.go]: should it be `!checkTableSame(region, adjacent)`?

File: server/schedule/checker/merge_checker.go [hunk 7]
```diff
@@ -134,8 +132,49 @@ func (m *MergeChecker) Check(region *core.RegionInfo) []*operator.Operator {
...
 }
+
+// allowMerge returns true if two regions can be merged according to the merge strategy.
+func (m *MergeChecker) allowMerge(region *core.RegionInfo, adjacent *core.RegionInfo) bool {
+	strategy := m.cluster.GetMergeStrategy()
+	switch strategy {
+	case "table":
+		allow := m.cluster.IsCrossTableMergeEnabled()
+		if allow {
+			return checkTable(region, adjacent)
```
Comment (disksing) [inline server/schedule/checker/merge_checker.go]: I think can return true directly.

File: server/statistics/metrics.go [hunk 8]
```diff
@@ -30,7 +30,7 @@ var (
 			Subsystem: "scheduler",
 			Name:      "store_status",
 			Help:      "Store status for schedule",
-		}, []string{"namespace", "address", "store", "type"})
+		}, []string{"address", "store", "type"})
```
Comment (disksing) [inline server/statistics/metrics.go]: Will it cause compatibility problem when upgrade from old version?

File: tools/pd-ctl/README.md [hunk 9]
```diff
@@ -203,6 +197,12 @@ Usage:
...
```
Comment (lhy1024) [inline tools/pd-ctl/README.md]: Do we need replace `enable-global-merge` with `enable-cross-table-merge`?

File: server/schedule/checker/merge_checker.go [hunk 10]
```diff
@@ -134,8 +132,49 @@ func (m *MergeChecker) Check(region *core.RegionInfo) []*operator.Operator {
...
+	default:
+		return checkTableSame(region, adjacent)
+	}
+}
+
+func checkTable(region *core.RegionInfo, adjacent *core.RegionInfo) bool {
+	return codec.Key(region.GetStartKey()).TableID() != 0 && codec.Key(adjacent.GetStartKey()).TableID() != 0
+}
+
+func checkTableSame(region *core.RegionInfo, adjacent *core.RegionInfo) bool {
```
Comment (lhy1024) [inline server/schedule/checker/merge_checker.go]:
```suggestion
func tableIsSame(region *core.RegionInfo, adjacent *core.RegionInfo) bool {
```

File: server/schedule/checker/merge_checker.go (hunk 6 above)
Comment (lhy1024) [inline server/schedule/checker/merge_checker.go]: we also need `checkTable`?

File: server/config/config.go (hunk 4 above)
Comment (rleungx) [reply server/config/config.go]: I don't think the "table" belongs to `cluster-type`. Maybe we need a better classification? But we cannot find a good one.

File: conf/config.toml (hunk 1 above)
Comment (rleungx) [reply conf/config.toml]: In fact, we need to replace all "region" with "Region"

File: server/statistics/metrics.go (hunk 8 above)
Comment (rleungx) [reply server/statistics/metrics.go]: Yes, it will. The monitor should also be upgraded under this situation.

Comment (disksing) [reply server/statistics/metrics.go]: Will old data be lost?

File: server/config/config.go (hunk 4 above)
Comment (disksing) [reply server/config/config.go]: or maybe `cluster-key-type`? I think it should not be restricted to merge feature. Other feature relies on it too.

File: server/schedule/checker/merge_checker.go [hunk 11]
```diff
@@ -134,8 +132,33 @@ func (m *MergeChecker) Check(region *core.RegionInfo) []*operator.Operator {
...
 		len(adjacent.GetPeers()) == m.cluster.GetMaxReplicas() // peer count should equal
 }
+
+// allowMerge returns true if two regions can be merged according to the merge strategy.
+func (m *MergeChecker) allowMerge(region *core.RegionInfo, adjacent *core.RegionInfo) bool {
+	strategy := m.cluster.GetMergeStrategy()
+	switch strategy {
+	case core.Table:
+		allow := m.cluster.IsCrossTableMergeEnabled()
+		if allow {
```
Comment (lhy1024) [inline server/schedule/checker/merge_checker.go]: I prefer to use `if m.cluster.IsCrossTableMergeEnabled()`

File: server/core/resource_kind.go [hunk 12]
```diff
@@ -90,3 +90,42 @@ func StringToScheduleStrategy(input string) ScheduleStrategy {
 		panic("invalid schedule strategy: " + input)
 	}
 }
+
+// MergeStrategy distinguishes different kinds of merge strategy
```
Comment (lhy1024) [inline server/core/resource_kind.go]: do we need to rename the file?

File: server/config/config.go (hunk 4 above)
Comment (rleungx) [reply server/config/config.go]: How about moving it into `PDServerConfig`?

File: server/statistics/metrics.go (hunk 8 above)
Comment (rleungx) [reply server/statistics/metrics.go]: Yes. It will. Any better idea?

File: server/core/resource_kind.go (hunk 12 above)
Comment (rleungx) [reply server/core/resource_kind.go]: Do you have a suggestion?

Comment (lhy1024) [reply server/core/resource_kind.go]: how about `kind.go`

File: server/config/config.go (hunk 4 above)
Comment (disksing) [reply server/config/config.go]: I think it's a good idea.

File: server/statistics/metrics.go (hunk 8 above)
Comment (disksing) [reply server/statistics/metrics.go]: seems no better idea. :(

File: server/config/config.go [hunk 13]
```diff
@@ -834,6 +791,9 @@ type PDServerConfig struct {
 	UseRegionStorage bool `toml:"use-region-storage" json:"use-region-storage,string"`
 	// MaxResetTSGap is the max gap to reset the tso.
 	MaxResetTSGap time.Duration `toml:"max-reset-ts-gap" json:"max-reset-ts-gap"`
+	// KeyType is option to specify the type of keys.
+	// There are some types supported: ["table", "raw", "txn"], default: "table"
+	KeyType string `toml:"key-type" json:"key-type"`
```
Comment (lhy1024) [inline server/config/config.go]: IMO, mergeStrategy is better than keyType

File: server/api/store.go [hunk 14]
```diff
@@ -81,7 +81,7 @@ func newStoreInfo(opt *config.ScheduleConfig, store *core.StoreInfo) *StoreInfo
 			Available:          typeutil.ByteSize(store.GetAvailable()),
 			LeaderCount:        store.GetLeaderCount(),
 			LeaderWeight:       store.GetLeaderWeight(),
-			LeaderScore:        store.LeaderScore(opt.GetLeaderScheduleStrategy(), 0),
+			LeaderScore:        store.LeaderScore(core.StringToScheduleStrategy(opt.LeaderScheduleStrategy), 0),
```
Comment (lhy1024) [inline server/api/store.go]: why change it

File: server/schedule/checker/merge_checker.go [hunk 15]
```diff
@@ -133,8 +131,32 @@ func (m *MergeChecker) Check(region *core.RegionInfo) []*operator.Operator {
...
+		return true
+	case core.Txn:
+		return true
+	default:
+		return isTableIDSame(region, adjacent)
+	}
+}
+
+func isTableIDSame(region *core.RegionInfo, adjacent *core.RegionInfo) bool {
+	if codec.Key(region.GetStartKey()).TableID() == codec.Key(adjacent.GetStartKey()).TableID() {
```
[the remark this answers is not in the document]
Comment (disksing) [reply server/schedule/checker/merge_checker.go]: LGTM except for this one.

**Comment by lhy1024** [comment]
could we close #1416 ?

**Pull request merged**